packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -100,29 +120,120 @@ export class SerializedStateManager {
 		supportGetSnapshotApi: boolean,
 	) {
 		if (this.pendingLocalState === undefined) {
-			const { snapshot, version } = supportGetSnapshotApi
-				? await fetchISnapshot(this.mc, this.storageAdapter, specifiedVersion)
-				: await fetchISnapshotTree(this.mc, this.storageAdapter, specifiedVersion);
-			const baseSnapshot: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
-				? snapshot.snapshotTree
-				: snapshot;
-			assert(baseSnapshot !== undefined, 0x8e4 /* Snapshot should exist */);
+			const { snapshotTree, version } = await getSnapshotTree(
+				this.mc,
+				this.storageAdapter,
+				supportGetSnapshotApi,
+				specifiedVersion,
+			);
 			// non-interactive clients will not have any pending state we want to save
 			if (this.offlineLoadEnabled) {
 				const snapshotBlobs = await getBlobContentsFromTree(
-					baseSnapshot,
+					snapshotTree,
 					this.storageAdapter,
 				);
-				this.snapshot = { baseSnapshot, snapshotBlobs };
+				this.snapshot = { baseSnapshot: snapshotTree, snapshotBlobs };
 			}
-			return { baseSnapshot, version };
+			return { baseSnapshot: snapshotTree, version };
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi)
+				.catch((error) => {
+					this.mc.logger.sendErrorEvent(
+						{
+							eventName: "Could_Not_Refresh_Snapshot",
+						},
+						error,
+					);
+					throw error;
+				})
+				.finally(() => {
+					this.refreshSnapshotLock = false;
+				});
 			return { baseSnapshot, version: undefined };
 		}
 	}
 
+	/**
+	 * Fetch latest snapshot available in storage and refresh class snapshot and processedOps
+	 * @param supportGetSnapshotApi -
+	 */
+	private async refreshSnapshot(supportGetSnapshotApi: boolean) {
+		if (this.refreshSnapshotLock) {
+			// refreshSnapshot is already being executed. Ignoring this call
+			return;
+		}
+		this.refreshSnapshotLock = true;
+		const { snapshotTree } = await getSnapshotTree(
+			this.mc,
+			this.storageAdapter,
+			supportGetSnapshotApi,
+			undefined,
+		);
+		await this.refreshPendingAttributes(snapshotTree);
+	}
+
+	/**
+	 * Updates class snapshot and processedOps with the latest fetched snapshot
+	 *
+	 * @param baseSnapshot -
+	 */
+	private async refreshPendingAttributes(baseSnapshot: ISnapshotTree) {
```

Comment:
i wouldn't split these functions. they are not useable independently

Reply:
You mean refreshPendingAttributes and refreshSnapshot? They are being used independently. I had the logic together but in case we're behind the snapshot seq num we don't need to kick off a new refresh but just use the snapshot we have already and just wait to catch up and just refresh the class attributes

Reply:
If we merge both functions I guess we couldn't use the LazyPromise approach since they would need to be called twice

Reply:
i only see one callsite for each function. where are the other calls?